File: Second_Year_Lab/OS_LabCode/Scheduling_Algorithm/roundRobin.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;

class Node{
    int p_id;
    int at;
    int bt;
    int Gcom=0;
    int tat;
    int wt;
    int rt;
    int dbt;
    int intime=-1;

    Node(int p_id,int at,int bt){
        this.p_id=p_id;
        this.at=at;
        this.bt=bt;
        this.dbt=bt;
    }
}

public class roundRobin {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of processes: ");
        int n = sc.nextInt();

        Node[] p = new Node[n];

        for (int i = 0; i < n; i++) {
            System.out.print("Enter PID AT BT: ");
            int pid = sc.nextInt();
            int at = sc.nextInt();
            int bt = sc.nextInt();
            p[i] = new Node(pid, at, bt);
        }

        System.out.println("Enter the quantam time: ");
        int qt=sc.nextInt();


        // Step 1: Sort by Arrival Time
        Arrays.sort(p, Comparator.comparingInt(a -> a.at));

        
        Queue<Node> q=new LinkedList<>();

        int completed=0;
        int i=0;
        Node currNode=p[0];
        while (!q.isEmpty() || i<n ) {
             if(!q.isEmpty()){
            currNode=q.poll();
         }else if(i<n){
            currNode=p[i];
            i++;

            if(completed < currNode.at)
                    completed = currNode.at;
         }

         if(currNode.intime==-1){
            currNode.intime=completed;
         }

         if(currNode.dbt>=qt)
            completed+=qt;
        else
            completed+=currNode.dbt;

         if(currNode.dbt>=qt)
            currNode.dbt-=qt;
        else    
            currNode.dbt=0;

        currNode.Gcom=completed;
        currNode.tat=currNode.Gcom-currNode.at;
        currNode.wt=currNode.tat-currNode.bt;
        currNode.rt=currNode.intime-currNode.at;
         

          // add newly arrived processes
            while(i<n && p[i].at<=completed){
                q.add(p[i]);
                i++;
            }

            // if process still has burst time
            if(currNode.dbt>0){
                q.add(currNode);
            }
        
         

        }

           Arrays.sort(p, Comparator.comparingInt(a -> a.p_id));

            for(int j=0;j<n;j++){
                Node curNode=p[j];
                 System.out.println(
                    "PID: " + curNode.p_id +
                    " AT: " + curNode.at +
                    " BT: " + curNode.bt +
                    " CT: " + curNode.Gcom +
                    " TAT: " + curNode.tat +
                    " WT: " + curNode.wt +
                    " RT: " + curNode.rt
            );
            }
    }
}

File: Second_Year_Lab/OS_LabCode/Scheduling_Algorithm/roundRobin.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int p_id;
    int at;
    int bt;
    int Gcom;
    int tat;
    int wt;
    int rt;
    int dbt;
    int intime;
};

void sortByArrival(struct Node p[], int n){
    for(int i=0;i<n-1;i++){
        for(int j=i+1;j<n;j++){
            if(p[i].at > p[j].at){
                struct Node temp = p[i];
                p[i] = p[j];
                p[j] = temp;
            }
        }
    }
}

void sortByPID(struct Node p[], int n){
    for(int i=0;i<n-1;i++){
        for(int j=i+1;j<n;j++){
            if(p[i].p_id > p[j].p_id){
                struct Node temp = p[i];
                p[i] = p[j];
                p[j] = temp;
            }
        }
    }
}

int main(){

    int n;
    printf("Enter number of processes: ");
    scanf("%d",&n);

    struct Node p[n];

    for(int i=0;i<n;i++){
        printf("Enter PID AT BT: ");
        scanf("%d %d %d",&p[i].p_id,&p[i].at,&p[i].bt);

        p[i].dbt = p[i].bt;
        p[i].intime = -1;
        p[i].Gcom = 0;
        p[i].tat = 0;
        p[i].wt = 0;
        p[i].rt = 0;
    }

    int qt;
    printf("Enter quantum time: ");
    scanf("%d",&qt);

    sortByArrival(p,n);

    int q[100];
    int front=0,rear=0;

    int completed=0;
    int i=0;
    int currIndex;

    while(front!=rear || i<n){

        if(front!=rear){
            currIndex = q[front];
            front++;
        }
        else{
            currIndex = i;
            i++;
            if(completed < p[currIndex].at)
                completed = p[currIndex].at;
        }

        if(p[currIndex].intime==-1)
            p[currIndex].intime = completed;

        int exec;

        if(p[currIndex].dbt >= qt)
            exec = qt;
        else
            exec = p[currIndex].dbt;

        p[currIndex].dbt -= exec;
        completed += exec;

        p[currIndex].Gcom = completed;
        p[currIndex].tat = p[currIndex].Gcom - p[currIndex].at;
        p[currIndex].wt = p[currIndex].tat - p[currIndex].bt;
        p[currIndex].rt = p[currIndex].intime - p[currIndex].at;

        while(i<n && p[i].at<=completed){
            q[rear] = i;
            rear++;
            i++;
        }

        if(p[currIndex].dbt > 0){
            q[rear] = currIndex;
            rear++;
        }
    }

    sortByPID(p,n);

    for(int j=0;j<n;j++){
        printf("PID: %d AT: %d BT: %d CT: %d TAT: %d WT: %d RT: %d\n",
        p[j].p_id,
        p[j].at,
        p[j].bt,
        p[j].Gcom,
        p[j].tat,
        p[j].wt,
        p[j].rt);
    }

    return 0;
}